c++/src/ColumnWriter.cc

```diff
@@ -0,0 +1,507 @@
+  }
+
+  void ColumnWriter::mergeStripeStatsIntoFileStats() {
+    colFileStatistics->merge(*colStripeStatistics);
+    colStripeStatistics->reset();
+  }
+
+  void ColumnWriter::mergeRowGroupStatsIntoStripeStats() {
+    colStripeStatistics->merge(*colIndexStatistics);
+    colIndexStatistics->reset();
+  }
+
+  void ColumnWriter::getFileStatistics(
+    std::vector<proto::ColumnStatistics>& stats) const {
+    getProtoBufStatistics(stats, colFileStatistics.get());
+  }
+
+  void ColumnWriter::createRowIndexEntry() {
+    proto::ColumnStatistics *indexStats = rowIndexEntry->mutable_statistics();
+    colIndexStatistics->toProtoBuf(*indexStats);
+
+    *rowIndex->add_entry() = *rowIndexEntry;
+
+    rowIndexEntry->clear_positions();
+    rowIndexEntry->clear_statistics();
+
+    if (enableStats) {
+      colStripeStatistics->merge(*colIndexStatistics);
+    }
+    colIndexStatistics->reset();
+
+    recordPosition();
+  }
+
+  void ColumnWriter::writeIndex(std::vector<proto::Stream> &streams) const {
+    // write row index to output stream
+    rowIndex->SerializeToZeroCopyStream(indexStream.get());
+
+    // construct row index stream
+    proto::Stream stream;
+    stream.set_kind(proto::Stream_Kind_ROW_INDEX);
+    stream.set_column(static_cast<uint32_t>(columnId));
+    stream.set_length(indexStream->flush());
+    streams.push_back(stream);
+  }
+
+  void ColumnWriter::recordPosition() const {
+    notNullEncoder->recordPosition(rowIndexPosition.get());
+  }
+
+  void ColumnWriter::resetIndex() {
+    // clear row index
+    rowIndex->clear_entry();
+    rowIndexEntry->clear_positions();
+    rowIndexEntry->clear_statistics();
+
+    // write current positions
+    recordPosition();
+  }
+
+  class StructColumnWriter : public ColumnWriter {
+  public:
+    StructColumnWriter(
+                       const Type& type,
+                       StreamsFactory& factory,
+                       const WriterOptions& options);
+    ~StructColumnWriter();
+
+    virtual void add(ColumnVectorBatch& rowBatch,
+                     uint64_t offset,
+                     uint64_t numValues) override;
+
+    virtual void flush(std::vector<proto::Stream>& streams) override;
+
+    virtual uint64_t getEstimatedSize() const override;
+    virtual void getColumnEncoding(
+      std::vector<proto::ColumnEncoding>& encodings) const override;
+
+    virtual void getStripeStatistics(
+      std::vector<proto::ColumnStatistics>& stats) const override;
+
+    virtual void getFileStatistics(
+      std::vector<proto::ColumnStatistics>& stats) const override;
+
+    virtual void mergeStripeStatsIntoFileStats() override;
+
+    virtual void mergeRowGroupStatsIntoStripeStats() override;
+
+    virtual void createRowIndexEntry() override;
+
+    virtual void writeIndex(
+      std::vector<proto::Stream> &streams) const override;
+
+    virtual void resetIndex() override;
+
+  private:
+    std::vector<ColumnWriter *> children;
```

Comment:
std::unique_ptr can be redefined as auto_ptr when unique_ptr is not supported by the current version of STL. See Adaptor.hh.in -

`#ifndef ORC_CXX_HAS_UNIQUE_PTR
  #define unique_ptr auto_ptr
#endif
`
And auto_ptr is not supposed to be used in STL container.